crates/handler/src/validation.rs

```diff
@@ -236,144 +266,110 @@ pub fn validate_tx_env<EvmWiringT: EvmWiring, SPEC: Spec>(
 {
-    let tx_type = tx.tx_type().into();
+    let tx_type = context.tx().tx_type().into();
     // EIP-3607: Reject transactions from senders with deployed code
     // This EIP is introduced after london but there was no collision in past
     // so we can leave it enabled always
-    if !cfg.is_eip3607_disabled() {
+    if !context.cfg().is_eip3607_disabled() {
         let bytecode = &account.info.code.as_ref().unwrap();
         // allow EOAs whose code is a valid delegation designation,
         // i.e. 0xef0100 || address, to continue to originate transactions.
         if !bytecode.is_empty() && !bytecode.is_eip7702() {
-            return Err(InvalidTransaction::RejectCallerWithCode);
+            return Err(InvalidTransaction::RejectCallerWithCode.into());
         }
     }
 
     // Check that the transaction's nonce is correct
-    if !cfg.is_nonce_check_disabled() {
-        let tx = tx.common_fields().nonce();
+    if !context.cfg().is_nonce_check_disabled() {
+        let tx = context.tx().common_fields().nonce();
         let state = account.info.nonce;
         match tx.cmp(&state) {
             Ordering::Greater => {
-                return Err(InvalidTransaction::NonceTooHigh { tx, state });
+                return Err(InvalidTransaction::NonceTooHigh { tx, state }.into());
             }
             Ordering::Less => {
-                return Err(InvalidTransaction::NonceTooLow { tx, state });
+                return Err(InvalidTransaction::NonceTooLow { tx, state }.into());
             }
             _ => {}
         }
     }
 
     // gas_limit * max_fee + value
-    let mut balance_check = U256::from(tx.common_fields().gas_limit())
-        .checked_mul(U256::from(tx.max_fee()))
-        .and_then(|gas_cost| gas_cost.checked_add(tx.common_fields().value()))
+    let mut balance_check = U256::from(context.tx().common_fields().gas_limit())
+        .checked_mul(U256::from(context.tx().max_fee()))
+        .and_then(|gas_cost| gas_cost.checked_add(context.tx().common_fields().value()))
         .ok_or(InvalidTransaction::OverflowPaymentInTransaction)?;
 
     if tx_type == TransactionType::Eip4844 {
-        let tx = tx.eip4844();
-        // if the tx is not a blob tx, this will be None, so we add zero
+        let tx = context.tx().eip4844();
         let data_fee = tx.calc_max_data_fee();
         balance_check = balance_check
-            .checked_add(U256::from(data_fee))
+            .checked_add(data_fee)
             .ok_or(InvalidTransaction::OverflowPaymentInTransaction)?;
     }
 
     // Check if account has enough balance for `gas_limit * max_fee`` and value transfer.
     // Transfer will be done inside `*_inner` functions.
-    if balance_check > account.info.balance {
-        if cfg.is_balance_check_disabled() {
-            // Add transaction cost to balance to ensure execution doesn't fail.
-            account.info.balance = account.info.balance.saturating_add(balance_check);
-        } else {
```

Comment:
This code was removed. I re-added it in https://github.com/bluealloy/revm/pull/2286